### PR TITLE
#168781675 - Signup/Register feature

### PR DESCRIPTION
**Pull request by nignanthomas**
#### What does this PR do?
Create Signup/Register Feature

#### Description of Task to be completed?
- Create integration tests to test the signup endpoint
- Edit the models to add password and isVerified fields
- Create validators for signup request body
- Create helpers/utils for hashPassord, generate JWT token, verifyEmail and verifyUsername
- Create signupController and export to routes
- Create midllewares for checkSignup for the req.body, verifyEmail, verifyUsername, confirmPassword
- Create routes for 'api/v1/register/' with the aforementioned middlewares and signupController
- Change node version in Dockerfile from node:11-alpine to node:11 due to bcrypt crash

#### How should this be manually tested?
- Setup the Application first (refer to README.md)
- Run `npm run dev`
- Download Postman.
- In Postman, create a new POST request tab and enter the following URL: `http://localhost:3000/api/v1/users/register`.
- Enter the following user object as raw/json under `Body` Section: 
```
{
	"email": "johndoe@gmail.com",
	"username": "johndoe08",
	"password": "Pa55w0rd",
	"confirmPassword": "Pa55w0rd"
}
```
- You shoud have s similar response to the following:  
```
{
    "message": "User Created Successfully",
    "data": {
        "user_id": 1,
        "username": "johndoe08",
        "email": "johndoe@gmail.com",
        "created_at": "Tue Oct 08 2019 14:45:14 GMT+0200 (Central Africa Time)",
        "updated_at": "Tue Oct 08 2019 14:45:14 GMT+0200 (Central Africa Time)",
        "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJwYXlsb2FkIjp7ImlkIjozfSwiaWF0IjoxNTcwNTM4NzE0LCJleHAiOjE1NzA2MjUxMTR9.2Xvj-XNmqouAhdSwUQviHz_QcrC7gxaWAU2vNZ9G_ps"
    }
}
```

#### Any background context you want to provide?
N/A

#### What are the relevant pivotal tracker stories?
[#168781675](https://www.pivotaltracker.com/story/show/168781675)

#### Screenshots (if appropriate)
Screen Shot 2019-10-08 at 15 26 52

#### Questions:
N/A

### CODE REVIEW

**Review by Cheza-Dzabala — CHANGES_REQUESTED** [review]
Review the codeclimate warning please

**Comment by nignanthomas** [comment]
> Review the codeclimate warning please

@Cheza-Dzabala Will do. Thanks
In the meantime, also check out the rest.

**Edit:** @Cheza-Dzabala codeclimate fixed => All checks passed, kindly review

**Review by alainmateso — CHANGES_REQUESTED** [review]
Great Job Thomas. However, when returning validation errors to the user, please make sure all validation errors are returned at once instead of returning one by one(In the screenshot you can see that it only returns an error for the password, yet confirm password is also wrong).  I also hope you don't mind making them user-friendly and more descriptive. I think the first part of the message would be enough. Regards
Screen Shot 2019-10-09 at 17 53 58

File: src/controllers/signupController.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import models from '../database/models';
+import responseUtil from '../utils/responseUtil';
+import strings from '../utils/stringsUtil';
+import hashPassword from '../utils/hashPassword';
+import generateToken from '../utils/generateToken';
+
+
```
Comment (nakiwuge) [inline src/controllers/signupController.js]: Remove this line

File: src/controllers/signupController.js [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+export default class SignupController {
+  static async signupController({ body: { username, email, password } }, res) {
```
Comment (nakiwuge) [inline src/controllers/signupController.js]: why are you using `async` and `static`

File: src/middlewares/checkSignup.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import validateSignup from '../validation/validateSignup';
+import responseUtil from '../utils/responseUtil';
+import strings from '../utils/stringsUtil';
+
+export default function checkSignup({ body }, res, next) {
```
Comment (nakiwuge) [inline src/middlewares/checkSignup.js]: since we agreed to use es6 use arrow function

File: src/middlewares/confirmPassword.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
+import responseUtil from '../utils/responseUtil';
+import strings from '../utils/stringsUtil';
+
+export default function confirmPassword({ body: { password, confirmPassword } }, res, next) {
```
Comment (nakiwuge) [inline src/middlewares/confirmPassword.js]: since we agreed to use es6 use arrow function

File: src/middlewares/verifyExist.js [hunk 5]
```diff
@@ -0,0 +1,8 @@
+import strings from '../utils/stringsUtil';
+import verifyExist from '../utils/verifyUserExist';
+
+export default function verifyUserExist({ body: { username, email } }, res, next) {
```
Comment (nakiwuge) [inline src/middlewares/verifyExist.js]: since we agreed to use es6 use arrow function

File: src/utils/verifyUserExist.js [hunk 6]
```diff
@@ -0,0 +1,12 @@
+import models from '../database/models';
+import responseUtil from './responseUtil';
+
+export default function verifyUserExist(res, property, message) {
```
Comment (nakiwuge) [inline src/utils/verifyUserExist.js]: since we agreed to use es6 use arrow function

File: src/validation/validateSignup.js [hunk 7]
```diff
@@ -0,0 +1,11 @@
+import Joi from '@hapi/joi';
+
+export default function validateSignup(data) {
```
Comment (nakiwuge) [inline src/validation/validateSignup.js]: since we agreed to use es6 use arrow function

**Review by nakiwuge — CHANGES_REQUESTED** [review]
@nignanthomas good job on the implementation. I have tested the feature and there is some feedback I would like you to work on.
- For security purposes, the password should contain at least 1  special character` #,@,?`
- The error returned to the user when the password is non-alphanumeric should be descriptive enough for the user...something like `Password should contain at least 1 letter,1 number and a special character`
- When the password or confirm password field is missing  a user should get an error `password required` or `confirmPasword required`
- The status code returned when a user successfully registers should be 201(created) not 200

I have also left comments on this pr

**Review by Cheza-Dzabala — APPROVED** [review]

**Review by alainmateso — APPROVED** [review]
Great Job @nignanthomas

**Review by sabin18 — APPROVED** [review]
It looks Good.

**Review by ngireric123 — APPROVED** [review]

**Review by nakiwuge — APPROVED** [review]

**Pull request merged**